### PR TITLE
[Common][PyTorch] Fused `apply_rotorary_pos_emb`

### PR DESCRIPTION
**Pull request by yaox12**
This PR contains a fused `apply_rotary_pos_emb` implementation, which fuses the sin/cos calculation, the rotation, the elementwise multiplies, and the concats into one single kernel. 

It supports `sbhd`, `bshd`, and `thd` formats, and the output tensor will be contiguous.

### CODE REVIEW

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -745,6 +745,7 @@ class RotaryPositionEmbedding(torch.nn.Module):
     def __init__(
         self,
         dim: int,
+        rotary_percent: float,
```
[the remark this answers is not in the document]
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:748]: `rotary_percent` is used in Megatron-LM/Core in https://github.com/NVIDIA/Megatron-LM/blob/443ce9f3f98fdc5a53c6b480c6e21b79944d198e/megatron/core/models/common/embeddings/rotary_pos_embedding.py#L36. 

While NeMo uses a lightly different `rotary_percentage` in its config and translates it to MCore's `rotary_percent` in https://github.com/NVIDIA/NeMo/blob/98848b00b323a8a67972e0c053e43e60e83b6bb3/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L317

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -816,10 +820,36 @@ def apply_rotary_pos_emb(t: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
...
+
+        cos_, sin_ = ctx.saved_tensors
+        grad_q = tex.fused_rope_backward(grad_output, cos_, sin_)
+
+        return grad_q, None, None
+
+def fused_apply_rotary_pos_emb(t: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
+    cos_ = torch.cos(freqs).to(t.dtype)
+    sin_ = torch.sin(freqs).to(t.dtype)
+    return FusedRoPEFunc.apply(t, cos_, sin_)
```
[the remark this answers is not in the document]
Comment (yaox12) [reply transformer_engine/pytorch/attention.py]: Changed accordingly. I made `fused` default to `False` because I'm not sure if it will break something, for example, ts or onnx things.

**Review by timmoon10 — APPROVED** [review]
That 4x speedup is quite nice. LGTM, pending CI.

**Comment by yaox12** [comment]
@timmoon10 Pushed a commit to address ci issues. Please trigger CI again.

**Comment by yaox12** [comment]
Realized that `test_fused_rope` was not added into QA. 
And it seems that the Jax/Paddle failures were not relevant to this PR? cc @timmoon10

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -797,6 +810,54 @@ def forward(self, max_seq_len: int, offset: int = 0):
         # emb [seq_length, .., dim]
         return emb.reshape(emb.size(0), 1, 1, emb.size(1))
 
+
+class FusedRoPEFunc(torch.autograd.Function):
+    """
+    Function for FusedRoPE
+
+    This implementation assumes the input tensor to be in `sbhd` or `thd` format and
+    the RoPE tensor to be of shape (s, 1, 1, d). It accepts arbitrary memory layouts to avoid
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention.py:819]: How difficult will it be to adapt this to a RoPE tensor `freq` of shape `(1, s, s, d)` and the input tensor `t` in `bshd` format? 
Basically, I wanna make RoPE work for both `sbhd` and `bshd` formats.

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -806,18 +867,68 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
+            "sbhd",
+            "thd",
+        ), f"Only 'sbhd' and 'thd' formats are supported when fused is True, got {qkv_format}."
+        assert not (
+            transpose_output_memory and qkv_format == "thd"
+        ), "transpose_output_memory is only supported with 'sbhd' format."
+        assert (
+            qkv_format == "sbhd" or cu_seqlens is not None
+        ), "cu_seqlens must not be None when qkv_format is 'thd'."
+        return FusedRoPEFunc.apply(
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention.py:911]: Can this function take in input batches of different sequence lengths? (like batch1 is `b, 256, h, d` while batch2 is `b, 240, s, d`). 
Something like this: https://github.com/NVIDIA/TransformerEngine/pull/557/files#diff-0af6d715a51b3efcebd6067805b5d17b64d25ef84399e256bade01a602ce4192R861-R873

I guess we could convert the input to `thd` format for such use-case but I'm wondering if it could still work without doing it?

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:819]: For `bshd` format, we could transpose the input tensor while keeping `freqs` of shape `(s, 1, 1, d)`. I'll push a version for your review.

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:911]: Is it a common case? `freqs` is usually calculated by taking a `max_seq_len` argument here. https://github.com/NVIDIA/TransformerEngine/blob/82555b3fd1ddece3a886fac3c7a4aa8d6f8c1955/transformer_engine/pytorch/attention.py#L809
So in most case, they should have the same seq length I think?

Of course I could add the support if needed.

Comment (yaox12) [reply transformer_engine/pytorch/attention.py:956]: Added it.

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py:859]: I think input tensor might be bigger than the `freqs` tensor, so transposing the former might incur some performance cost

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention.py:956]: Ah I didn't know the intended use of the `forward` method in every iteration. Thanks for pointing it out!
I still feel that if the `max_seq_len` is globally the max, then we could simply get slice from this tensor for every smaller batch we get, instead of creating it afresh in every iteration.

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:859]: `transpose` just changes the layout of a tensor but won't do the real transposing in memory. And luckily, this fused rope kernel accepts arbitrary layouts, so ideally it won't introduce any extra cost.

File: transformer_engine/pytorch/attention.py [hunk 5]
```diff
@@ -1034,41 +1103,79 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
-                                        "upto {max_seq_len} sequence length!")
-    freqs = freqs[:cur_seq_len].to(t.dtype)
+    assert cur_seq_len <= max_seq_len, (
+        f"Rotary Embeddings only supported up to {max_seq_len} sequence length!"
+    )
+    freqs = freqs[:cur_seq_len]
     if tensor_format == "bshd":
-        freqs = freqs.transpose(0,1) # [seq, 1, 1, dim] -> [1, seq, 1, dim]
+        freqs = freqs.transpose(0, 1)  # [seq, 1, 1, dim] -> [1, seq, 1, dim]
+    cos_ = torch.cos(freqs).to(t.dtype)
```
Comment (yaox12) [inline transformer_engine/pytorch/attention.py:1169]: Calculate cos/sin first for better precision.

**Comment by yaox12** [comment]
@timmoon10 @sudhakarsingh27 Can I have another round of review? I merged PR #557 and checked via nsys that kernels are well fused for `sbhd`, `bshd` and `thd` formats.

File: transformer_engine/pytorch/attention.py (hunk 3 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:1047]: Checked via nsys that there's no extra cost.

File: transformer_engine/pytorch/attention.py (hunk 4 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:1146]: It's supported now.

File: transformer_engine/pytorch/attention.py [hunk 6]
```diff
@@ -1031,23 +1040,58 @@ def forward(self, max_seq_len: int, offset: int = 0):
...
+    """
+
     @staticmethod
     def forward(
-        ctx, t: torch.Tensor, cos_: torch.Tensor, sin_: torch.Tensor
+        ctx,
+        t: torch.Tensor,
+        freqs: torch.Tensor,
+        transpose_output_memory: bool = False,
+        qkv_format: str = "sbhd",
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1057]: Typically we would expect the output format to match the input format, so the `transpose_output_memory` kwarg seems to make the API more complicated than necessary. Could we remove `transpose_output_memory` as a kwarg and internally calculate based on `qkv_format == "bshd"`?

File: transformer_engine/pytorch/attention.py [hunk 7]
```diff
@@ -1062,37 +1106,78 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
 def apply_rotary_pos_emb(
     t: torch.Tensor,
     freqs: torch.Tensor,
+    tensor_format: str = "sbhd",
     fused: bool = False,
+    transpose_output_memory: bool = False,
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1111]: Is `transpose_output_memory` kwarg needed?

File: transformer_engine/pytorch/attention.py (hunk 6 above)
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1057]: `tensor_format` is more descriptive than `qkv_format` and is consistent with `apply_rotary_pos_emb`:
```suggestion
        tensor_format: str = "sbhd",
```

File: transformer_engine/pytorch/attention.py [hunk 8]
```diff
@@ -1062,37 +1106,78 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
-        t = (t * cos_) + (_rotate_half(t) * sin_)
-        return torch.cat((t, t_pass), dim=-1)
+    rot_dim = freqs.shape[-1]
+    # ideally t_pass is empty so rotary pos embedding is applied to all tensor t
+    t, t_pass = t[..., :rot_dim], t[..., rot_dim:]
+
+    # first part is cosine component
+    # second part is sine component, need to change signs with _rotate_half method
+    t = (t * cos_) + (_rotate_half(t) * sin_)
+    return torch.cat((t, t_pass), dim=-1)
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1180]: To return a BSHD tensor:
```suggestion
    emb = torch.cat((t, t_pass), dim=-1)
    if tensor_format == "bshd":
        emb = emb.transpose(0, 1).contiguous()
    return emb
```

File: tests/pytorch/test_fused_rope.py [hunk 9]
```diff
@@ -1,38 +1,164 @@
...
+    return dict(atol=1e-5, rtol=1.3e-6)
+
+
+def _overlapping_grad(output: torch.Tensor) -> torch.Tensor:
+    return output.sum() * 2
+
+
+def _non_overlapping_grad(output: torch.Tensor) -> torch.Tensor:
+    t = torch.ones_like(output)
+    return torch.sum(output * t)
```
Comment (timmoon10) [inline tests/pytorch/test_fused_rope.py:50]: What is this loss function testing? Isn't it equivalent to `output.sum()`?

File: transformer_engine/pytorch/attention.py (hunk 7 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py]: Removed.

File: transformer_engine/pytorch/attention.py (hunk 6 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py]: Changed accordingly.

Comment (yaox12) [reply transformer_engine/pytorch/attention.py]: Removed `transpose_output_memory` and add a comment to explain the behavior.

File: transformer_engine/pytorch/attention.py (hunk 8 above)
Comment (yaox12) [reply transformer_engine/pytorch/attention.py:1166]: I think `tensor_format` is used to indicate the format of input tensor. When `tensor_format == "bshd"`, `t` should already be in `bshd` and `freqs` has been tranposed. So Line 1179 is done in `bshd`, and we don't need the transpose here.

File: tests/pytorch/test_fused_rope.py (hunk 9 above)
Comment (yaox12) [reply tests/pytorch/test_fused_rope.py:50]: `output.sum()` produces a scalar. In order to do backward calculation, it's broadcasted to a tensor. This is "overlapping" memory. In this `_non_overlapping_grad` loss function, the multiply `output * t` will receive an overlapping grad, then the fused RoPE function will receive a non-overlapping grad.

I added two different loss functions to make sure I were reading the grad tensor correctly.

File: transformer_engine/pytorch/attention.py (hunk 6 above)
Comment (timmoon10) [reply transformer_engine/pytorch/attention.py]: I don't think this is the right API since it still forces the output to be SBHD. It's unintuitive for the user if they put in a BSHD input.

File: transformer_engine/pytorch/attention.py [hunk 10]
```diff
@@ -1025,6 +1038,58 @@ def forward(self, max_seq_len: int, offset: int = 0):
...
+    @staticmethod
+    def forward(
+        ctx,
+        t: torch.Tensor,
+        freqs: torch.Tensor,
+        tensor_format: str = "sbhd",
+        cu_seqlens: Union[torch.Tensor, None] = None,
+    ) -> torch.Tensor:
+        if tensor_format == "sbhd":
+            output = tex.fused_rope_forward(t, freqs, True)
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1060]: We want the output tensor to be contiguous:
```suggestion
            output = tex.fused_rope_forward(t, freqs, False)
```

File: transformer_engine/pytorch/attention.py [hunk 11]
```diff
@@ -1025,6 +1038,58 @@ def forward(self, max_seq_len: int, offset: int = 0):
...
+
+        return output
+
+    @staticmethod
+    def backward(
+        ctx, grad_output: torch.Tensor
+    ) -> Tuple[Union[torch.Tensor, None], ...]:
+        freqs, cu_seqlens = ctx.saved_tensors
+        if ctx.tensor_format == "sbhd":
+            grad_input = tex.fused_rope_backward(grad_output, freqs, True)
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1080]:
```suggestion
            grad_input = tex.fused_rope_backward(grad_output, freqs, False)
```

File: transformer_engine/pytorch/attention.py [hunk 12]
```diff
@@ -1035,46 +1100,69 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
+        rotary positional embedding will be applied.
+    freqs: torch.Tensor
+        Rotary positional embedding tensor of shape `[s2, 1, 1, d2]` and dtype 'float',
+        with `s2 >= s` and `d2 <= d`.
+    fused: bool, default = False
+        Whether to use a fused applying RoPE implementation.
+        Note: When using fused RoPE on an input tensor in `sbhd` format, the output
+        tensor will be in shape `[s, b, h, d]` but the 's' and 'b' dimensions of the
+        underlying memory layout are transposed.
+        See https://github.com/NVIDIA/TransformerEngine/pull/517 for more details.
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:1125]: Why is the output not contiguous in the SBHD case? Your kernel should be flexible enough to handle contiguous outputs for both SBHD and BSHD.

File: tests/pytorch/test_fused_rope.py (hunk 9 above)
Comment (timmoon10) [reply tests/pytorch/test_fused_rope.py:50]: Ah, makes sense to me then. Can you add a comment to make it easier to understand in the future?

File: tests/pytorch/test_fused_rope.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    loss_fused.backward()
+    grad_fused = t.grad.detach().clone()
+    t.grad = None
+
+    torch.testing.assert_close(output_fused, output_unfused, **get_tol(dtype))
+    torch.testing.assert_close(grad_fused, grad_unfused, **get_tol(dtype))
+    if tensor_format == "bshd":
+        assert output_fused.is_contiguous()
+    else:
+        assert output_fused.transpose(0, 1).is_contiguous()
```
Comment (timmoon10) [inline tests/pytorch/test_fused_rope.py:113]: If we support contiguous outputs for both the SBHD and BSHD cases:
```suggestion
    assert output_fused.is_contiguous()
```

File: tests/pytorch/test_fused_rope.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    return dict(atol=1e-5, rtol=1.3e-6)
+
+
+def _overlapping_grad(output: torch.Tensor) -> torch.Tensor:
+    return output.sum() * 2
+
+
+def _non_overlapping_grad(output: torch.Tensor) -> torch.Tensor:
+    t = torch.ones_like(output)
+    return torch.sum(output * t)
```
Comment (timmoon10) [inline tests/pytorch/test_fused_rope.py:50]:
```suggestion
# Gradient is a broadcasted scalar
def _overlapping_grad(output: torch.Tensor) -> torch.Tensor:
    return output.sum() * 2

# Gradient is a full tensor
def _non_overlapping_grad(output: torch.Tensor) -> torch.Tensor:
    t = torch.ones_like(output)
    return torch.sum(output * t)
```

**Review by timmoon10 — APPROVED** [review]
I've made some changes so the outputs are contiguous in both the SBHD and BSHD cases. If there are no objections and no CI issues, I think this is good to go.

**Comment by yaox12** [comment]
> I've made some changes so the outputs are contiguous in both the SBHD and BSHD cases. If there are no objections and no CI issues, I think this is good to go.

I intended to avoid the later `x.transpose(0, 1).contiguous` in FlashAttention. https://github.com/NVIDIA/TransformerEngine/blob/5debfdb356f6a6caaab070bd7e3f4052bb49df58/transformer_engine/pytorch/attention.py#L1282

But your changes are fine to me. We can use `bshd` format in this case.

**Comment by timmoon10** [comment]
Ah, I see the reasoning now. I'm unsure the best approach to handle the fact that Flash Attention expects BSHD tensors while the unfused backend and cuDNN fused attention expect SBHD. Pinging @cyanguwa. In any case, I'll merge this and we can optimize away the transpose in the future if needed.

File: transformer_engine/pytorch/attention.py [hunk 15]
```diff
@@ -1035,46 +1100,65 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
...
+    cos_ = torch.cos(freqs).to(t.dtype)
+    sin_ = torch.sin(freqs).to(t.dtype)
 
     rot_dim = freqs.shape[-1]
     # ideally t_pass is empty so rotary pos embedding is applied to all tensor t
     t, t_pass = t[..., :rot_dim], t[..., rot_dim:]
 
     # first part is cosine component
     # second part is sine component, need to change signs with _rotate_half method
-    t = (t * freqs.cos()) + (_rotate_half(t) * freqs.sin())
```
Comment (yspMing) [inline transformer_engine/pytorch/attention.py:1077]: Before doing the last calculation , should we transform the input into odd and even parts?
like
```
s,b,h,d = t.shape
t = t.view(s, b, h, d // 2, 2).transpose(4, 3).reshape(s, b, h, d)
```
